### PR TITLE
ROS CI with code coverage reporting

### PR DESCRIPTION
**Pull request by tylerjw**
This configures the CI to build with various cmake presets, generate code coverage reports, and build using the ros-ci system.

### CODE REVIEW

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -6,34 +6,74 @@ on:
...
-    container: ghcr.io/picknikrobotics/rsl:upstream-${{ matrix.ROS_DISTRO }}
+    container: ghcr.io/picknikrobotics/rsl:upstream-${{ matrix.ros_distro }}
     env:
       CCACHE_DIR: ${{ github.workspace }}/.ccache
     steps:
       - name: Check out the repo
         uses: actions/checkout@v3
       - name: Cache ccache
-        uses: pat-s/always-upload-cache@v3.0.1
+        uses: pat-s/always-upload-cache@v3
```
Comment (ChrisThrasher) [inline .github/workflows/ci.yaml:22]: Does this mean we always get the latest release in the 3.x series?

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -6,34 +6,74 @@ on:
...
         with:
           path: ${{ env.CCACHE_DIR }}
-          key: ccache-${{ github.sha }}-${{ github.run_id }}
+          key: ccache-cmake-${{ matrix.ros_distro }}-${{ matrix.preset }}-${{ github.sha }}-${{ github.run_id }}
           restore-keys: |
-            ccache-${{ github.sha }}
-            ccache
+            ccache-cmake-${{ matrix.ros_distro }}-${{ matrix.preset }}-${{ github.sha }}
+            ccache-cmake-${{ matrix.ros_distro }}-${{ matrix.preset }}
```
[the remark this answers is not in the document]
Comment (tylerjw) [reply .github/workflows/ci.yaml:28]: It takes about 4min to build catch2 from scratch without this.

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (tylerjw) [reply .github/workflows/ci.yaml:22]: Yes, I didn't see any reason to specify the patch version here.

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -6,34 +6,74 @@ on:
...
+        if: ${{ matrix.preset == 'codecov' }}
+        run: lcov -c -d . -o coverage.info --no-external --exclude "*/build/**"
+      - uses: codecov/codecov-action@v3
+        if: ${{ matrix.preset == 'codecov' }}
+        with:
+          files: ./coverage.info
+          name: codecov-umbrella
+          flags: ${{ matrix.ros_distro }}
+
+  ros:
```
[the remark this answers is not in the document]
Comment (tylerjw) [reply .github/workflows/ci.yaml:51]: To learn about this ci pipeline and make sure we don't write anything that builds in cmake but not using colcon.

File: .github/workflows/ci.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (ChrisThrasher) [reply .github/workflows/ci.yaml:28]: Right, so is it feasible to measure how much time is being saved per CI run? For all we know it takes 4 minutes to upload and download the ccache cache.

**Review by ChrisThrasher — APPROVED** [review]

**Pull request merged**